iroh-bytes/src/store/file.rs

```diff
@@ -0,0 +1,2483 @@
+    /// Non short lived partial entries always live in the file system, and for
+    /// short lived ones we never create a database entry in the first place.
+    Partial {
+        /// Once we get the last chunk of a partial entry, we have validated
+        /// the size of the entry despite it still being incomplete.
+        ///
+        /// E.g. a giant file where we just requested the last chunk.
+        size: Option<u64>,
+    },
+}
+
+impl Default for EntryState {
+    fn default() -> Self {
+        Self::Partial { size: None }
+    }
+}
+
+impl EntryState {
+    fn union(self, that: Self) -> io::Result<Self> {
+        match (self, that) {
+            (
+                Self::Complete {
+                    data_location,
+                    outboard_location,
+                },
+                Self::Complete {
+                    data_location: b_data_location,
+                    ..
+                },
+            ) => Ok(Self::Complete {
+                // combine external paths if needed
+                data_location: data_location.union(b_data_location)?,
+                outboard_location,
+            }),
+            (a @ Self::Complete { .. }, Self::Partial { .. }) =>
+            // complete wins over partial
+            {
+                Ok(a)
+            }
+            (Self::Partial { .. }, b @ Self::Complete { .. }) =>
+            // complete wins over partial
+            {
+                Ok(b)
+            }
+            (Self::Partial { size: a_size }, Self::Partial { size: b_size }) =>
+            // keep known size from either entry
+            {
+                Ok(Self::Partial {
+                    size: a_size.or(b_size),
+                })
+            }
+        }
+    }
+}
+
+impl redb::RedbValue for EntryState {
+    type SelfType<'a> = EntryState;
+
+    type AsBytes<'a> = SmallVec<[u8; 128]>;
+
+    fn fixed_width() -> Option<usize> {
+        None
+    }
+
+    fn from_bytes<'a>(data: &'a [u8]) -> Self::SelfType<'a>
+    where
+        Self: 'a,
+    {
+        postcard::from_bytes(data).unwrap()
+    }
+
+    fn as_bytes<'a, 'b: 'a>(value: &'a Self::SelfType<'b>) -> Self::AsBytes<'a>
+    where
+        Self: 'a,
+        Self: 'b,
+    {
+        postcard::to_extend(value, SmallVec::new()).unwrap()
+    }
+
+    fn type_name() -> redb::TypeName {
+        redb::TypeName::new("EntryState")
+    }
+}
+
+/// Options for inlining small complete data or outboards.
+#[derive(Debug, Clone)]
+pub struct InlineOptions {
+    /// Maximum data size to inline.
+    max_data_inlined: u64,
+    /// Maximum outboard size to inline.
+    max_outboard_inlined: u64,
+}
+
+impl InlineOptions {
+    /// Do not inline anything, ever.
+    pub const NO_INLINE: Self = Self {
+        max_data_inlined: 0,
+        max_outboard_inlined: 0,
+    };
+    /// Always inline everything
+    pub const ALWAYS_INLINE: Self = Self {
+        max_data_inlined: u64::MAX,
```

Comment:
I heard someone didn't like sentinel values, so why not make this `Option<u64>` and in this case set it to `None`?

Reply:
Hm, this is inclusive, so setting this to u64::MAX actually means inlining everything including the unlikely u64::max sized blob. The downside is that you can't disable inlining the single possible blob that has a size of 0.